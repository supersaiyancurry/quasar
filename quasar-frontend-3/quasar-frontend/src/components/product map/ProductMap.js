import React, { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import LogRocket from 'logrocket';
import ErrorPage from '../error page/ErrorPage';
import ProductCard from '../product card/ProductCard';
import { QUASAR_API, ERROR } from '../Constants';
import styles from './ProductMap.module.css';

const ProductMap = () => {
  const [products, setProducts] = useState([]);
  const [errorCode, setErrorCode] = useState();
  const [isLoading, setIsLoading] = useState(false);
  const { demographic, category, type } = useParams();

  useEffect(() => {
    let url;
    if (type) {
      url = `${QUASAR_API}/products/demographic/${demographic}/category/${category}/type/${type}`;
    } else if (category) {
      url = `${QUASAR_API}/products/demographic/${demographic}/category/${category}`;
    } else if (demographic) {
      url = `${QUASAR_API}/products/demographic/${demographic}`;
    } else {
      url = `${QUASAR_API}/products`;
    }

    const fetchData = async () => {
      let serverError;
      setIsLoading(true);
      await fetch(url, {
        method: 'GET',
        headers: new Headers({
          'Content-Type': 'application/json'
        })
      })
        .then((response) => {
          if (response.ok) {
            return response.json();
          }
          serverError = response.status;
          throw new Error(<p>{ERROR}</p>);
        })
        .then((productData) => {
          setProducts(productData);
          setIsLoading(false);
        })
        .catch(() => {
          // if server doesn't respond with an error message, display a 503
          setErrorCode(serverError || 503);
          LogRocket.error(serverError || '503: the service is unavailable');
          // scrolls back to the top so user can see the error
          document.body.scrollTop = 0; // For Safari
          document.documentElement.scrollTop = 0; // For Chrome, Firefox, IE and Opera
          setIsLoading(false);
        });
    };
    fetchData();
  }, [category, demographic, type]);

  return (
    <div>
      <nav aria-label="breadcrumb">
        <ol className={`breadcrumb ${styles.breadcrumb}`}>
          <li className={`breadcrumb-item ${styles.home}`}><a className={styles.link} href="/">Home</a></li>
          {demographic && <li className={`breadcrumb-item ${styles.item}`}><a className={styles.link} href={`/products/${demographic}`}>{demographic}</a></li>}
          {category && <li className={`breadcrumb-item ${styles.item}`}><a className={styles.link} href={`/products/${demographic}/${category}`}>{category}</a></li>}
          {type && <li className={`breadcrumb-item ${styles.item}`}><a className={styles.link} href={`/products/${demographic}/${category}/${type}`}>{type}</a></li>}
        </ol>
      </nav>
      <h3 className={styles.title}>Product Results</h3>
      {errorCode && <ErrorPage errorCode={errorCode} />}
      {isLoading && (
        <div className={styles.loading}>
          <p className="text-warning">One moment...</p>
          {LogRocket.log('Server is loading')}
        </div>
      )}
      {!isLoading && !errorCode && products.length === 0 && (
        <div className="d-flex justify-content-center">
          <h1 className={styles.empty}>Sorry, nothing to see here...</h1>
          {LogRocket.error('No products were retrieved with this fetch call')}
        </div>
      )}
      <div className="row row-cols-1 row-cols-lg-2 row-cols-xl-3 m-4">
        {products && products.map((product) => (
          <div key={product.id}>
            <ProductCard product={product} key={product.id} />
          </div>
        ))}
      </div>
    </div>
  );
};

export default ProductMap;
