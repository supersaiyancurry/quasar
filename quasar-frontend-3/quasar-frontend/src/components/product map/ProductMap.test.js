import React from 'react';
import { createMemoryHistory } from 'history';
import fetchMock from 'jest-fetch-mock';
import { render, cleanup, waitFor } from '@testing-library/react';
import { Router } from 'react-router-dom';
import ProductMap from './ProductMap';

afterEach(cleanup);

fetchMock.enableMocks();

const products = [
  {
    id: 163,
    price: 50.71,
    name: 'Heavy Duty Soccer Hoodie',
    description: "This soccer hoodie for men is extremely heavy duty. You're going to love it!",
    demographic: 'Men',
    category: 'Soccer',
    type: 'Hoodie',
    releaseDate: '2017-01-01',
    primaryColorCode: '#3079ab',
    secondaryColorCode: '#f092b0',
    styleNumber: 'sc97427',
    globalProductCode: 'po-7560240'
  },
  {
    id: 3426,
    price: 70.36,
    name: 'Trendy Baseball Shin Guard',
    description: "This baseball shin guard for men is extremely trendy. You're going to love it!",
    demographic: 'Men',
    category: 'Baseball',
    type: 'Shin Guard',
    releaseDate: '2017-01-01',
    primaryColorCode: '#e15258',
    secondaryColorCode: '#000000',
    styleNumber: 'sc06765',
    globalProductCode: 'po-0223877'
  }
];

const renderWithRouter = (component) => {
  const history = createMemoryHistory();
  return {
    ...render(
      <Router history={history}>
        {component}
      </Router>
    )
  };
};

it('renders without crashing', () => {
  renderWithRouter(
    <ProductMap />
  );
});

it('product card component matches snapshot', () => {
  const { asFragment } = renderWithRouter(
    <ProductMap />
  );
  expect(asFragment()).toMatchSnapshot();
});

it('should load product data from api', async () => {
  fetch.mockResponse(JSON.stringify({ data: products }));
  const { getAllByTestId } = renderWithRouter(<ProductMap />);
  const productData = await waitFor(() => getAllByTestId('products'));

  expect(productData).toHaveLength(2);
});
