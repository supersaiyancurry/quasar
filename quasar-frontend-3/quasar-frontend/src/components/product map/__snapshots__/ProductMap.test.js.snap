// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`product card component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <nav
      aria-label="breadcrumb"
    >
      <ol
        class="breadcrumb breadcrumb"
      >
        <li
          class="breadcrumb-item home"
        >
          <a
            class="link"
            href="/"
          >
            Home
          </a>
        </li>
      </ol>
    </nav>
    <h3
      class="title"
    >
      Product Results
    </h3>
    <div
      class="loading"
    >
      <p
        class="text-warning"
      >
        One moment...
      </p>
    </div>
    <div
      class="row row-cols-1 row-cols-lg-2 row-cols-xl-3 m-4"
    />
  </div>
</DocumentFragment>
`;
